File: Sales_Billing/src/dplyr/merge_KE30_sales.R
library(dplyr)
library(readr)
library(purrr)
library(fs)
library(stringr)
library(janitor)
library(here)
library(glue)


# Path
path <- here("Sales_Billing")


# Functions
read_txt_file <- function(file_path) {
  # Read tsv file
  df <- read_tsv(file_path,
    skip = 7,
    locale = locale(encoding = "UTF-16LE"),
    col_types = cols(
      "RecordType" = col_character(),
      "Customer" = col_character(),
      "AAG" = col_character(),
      "Dv" = col_character(),
    ),
  )
  return(df)
}


read_multiple_files <- function(list_of_files) {
  df <- list_of_files |>
    map_dfr(read_txt_file)
  return(df)
}


remove_columns <- function(df) {
  df <- df |>
    select(!c("x1", "x4"))
  return(df)
}


remove_missing_values <- function(df) {
  df <- df |>
    filter(!is.na(df$record_type))
  return(df)
}


extract_profit_center <- function(df) {
  df <- df |>
    mutate(
      profit_center = str_replace(.data$profit_center, "5668/", "")
    )
  return(df)
}


main <- function() {
  # Variables
  year <- "2024"

  # Path
  # data_path <- here(path, "data", "KE30", "Archive", glue("{year}"))
  data_path <- here(path, "data", "KE30")  # current year

  # Filenames
  input_files <- dir_ls(data_path, regexp = "\\.XLS|\\.xls")
  output_file <- here(path, "db", glue("KE30_{year}.csv"))

  # Read data
  df <- read_multiple_files((input_files)) |>
    clean_names()

  # Process data
  df <- df |>
    remove_columns() |>
    remove_missing_values() |>
    extract_profit_center()

  # Write data
  write_csv(df, output_file, na = "")
  print("A file is created")
}

main()

File: Sales_Billing/src/pandas/merge_KE30_sales.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def read_multiple_files(list_of_files):
    dataframes = [
        pd.read_csv(
            file,
            skiprows=7,
            sep="\t",
            encoding="UTF-16LE",
            skipinitialspace=True,
            thousands=",",
            engine="python",
            dtype={
                # text
                "AAG": str,
                "Customer": str,
                "Tr.Prt": str,
                "Dv": str,
                "Dir Cust.": str,
                "Fin.Cust.": str,
                # number
                "Revenues": int,
                "Stock val.": int,
                "Quantity": int,
            },
        ).rename(
            # To avoid the same column names after clean_names
            columns={
                "Record Type": "Record Type 2",
            }
        )
        for file in list_of_files
    ]

    # Merge the list of DataFrames into a single DataFrame
    df = pd.concat(dataframes)

    return df


def remove_columns(df):
    cols_to_remove = ["unnamed_0", "unnamed_3"]
    df = df[[col for col in df.columns if col not in cols_to_remove]]
    return df


def remove_missing_values(df):
    df = df.dropna(subset=["record_type"])
    return df


def extract_profit_center(df):
    df["profit_center"] = df["profit_center"].str.replace("5668/", "")
    return df


def main():
    # Variables
    year = "2024"

    # Path
    # data_path = path / "data" / "KE30" / "Archive" / f"{year}"
    data_path = path / "data" / "KE30"  # for current year

    # Filenames
    input_files = [
        file
        for file in data_path.iterdir()
        if file.is_file() and (file.suffix.lower() == ".xls")
    ]
    output_file = path / "db" / f"KE30_{year}.csv"

    # Read data
    df = read_multiple_files(input_files).clean_names(
        strip_underscores=True, case_type="snake"
    )

    # Process data
    df = df.pipe(remove_columns).pipe(remove_missing_values).pipe(extract_profit_center)

    # Write data
    df.to_csv(output_file, index=False)
    print("A file is created")


if __name__ == "__main__":
    main()
